R/read.R

```diff
@@ -180,61 +180,92 @@ st_list = function(dsn, options = character(0), do_count = FALSE) {
-                       "csv",    "CSV",
-                       "e00",    "AVCE00",
-                       "gdb",    "FileGDB",
-                       "geojson","GeoJSON",
-                       "gml",    "GML",
-                       "gmt",    "GMT",
-                       "gpkg",   "GPKG",
-                       "gps",    "GPSBabel",
-                       "gtm",    "GPSTrackMaker",   
-                       "gxt",    "Geoconcept",
-                       "jml",    "JML",
-                       "map",    "WAsP",
-                       "mdb",    "Geomedia",
-                       "nc",     "netCDF",
-                       "ods",    "ODS",
-                       "osm",    "OSM",
-                       "pbf",    "OSM",
-                       "shp",    "ESRI Shapefile",
-                       "sqlite", "SQLite",
-                       "vdv",    "VDV",
-                       "xls",    "xls",
-                       "xlsx",   "XLSX"
-                     ), ncol = 2, byrow = TRUE)
-	prefix_map = matrix(c(
-                       "couchdb", "CouchDB",
-                       "DB2ODBC", "DB2ODBC",
-                       "DODS",    "DODS",
-                       "GFT",     "GFT",
-                       "MSSQL",   "MSSQLSpatial",
-                       "MySQL",   "MySQL",
-                       "OCI",     "OCI",
-                       "ODBC",    "ODBC",
-                       "PG",      "PostgreSQL",
-                       "SDE",     "SDE"
-                     ), ncol = 2, byrow = TRUE)
-	drv = ext_map[,2]
-	names(drv) = ext_map[,1]
-	prfx = prefix_map[,2]
-	names(prfx) = tolower(prefix_map[,1])
-
+  stopifnot(is.character(dsn))
+  stopifnot(length(dsn) == 1)
+  
 	# find match: try extension first
-	drv = drv[tolower(tools::file_ext(dsn))]
-	if (is.na(drv)) { # try prefix, like "PG:dbname=sth"
-		if (length(grep(":", dsn)))
-			drv = prfx[tolower(strsplit(dsn, ":")[[1]][1])]
-		if (is.na(drv)) # no match
-			stop("no driver specified, cannot guess driver from dsn")
+	drv = extension_map()[tolower(tools::file_ext(dsn))]
+	if (any(grep(":", dsn))) {
+			drv = prefix_map()[tolower(strsplit(dsn, ":")[[1]][1])]
 	}
-	drivers = st_drivers()
-	i = match(drv, drivers$name)
-	if (is.na(i))
-		stop(paste("guess_driver:", drv, "not available in supported drivers, see `st_drivers()'"))
-	if (! drivers[i, "write"])
-		stop(paste("Driver", drv, "has no write capability, see `st_drivers()'"))
-	drv
+	if (is.null(drv)) {
+	  # no match
+	  return(NA)
+	}
+	
+	unlist(drv)
+}
+
+guess_driver_can_write = function(dns, drv = guess_driver(dns)) {
+  if(is.na(drv)) {
+    stop("Could not guess driver for ", dns, call. = FALSE)
+  }
+  if(!is_driver_available(drv)) {
+    stop(paste(unlist(drv), " not available in supported drivers, see `st_drivers()'"), call. = FALSE)
+  }
+  if(!is_driver_can(drv, operation="write")) {
+    stop(paste("Driver ", drv, " cannot `", operation, "`, see `st_drivers()'"), call. = FALSE)
```

Comment:
I'll look into it, I agree on code coverage. R checks passes, at least locally, I'll look at the Travis log. I wasn't sure we could safely rely on minimal driver support, that's why there is no specific test for it. I could add a test relying on `shp` support since it is used implicitly to load data for different tests. I think it's a reasonable assumption, and no `shp` support would mean many other problems.